### PR TITLE
feat(optimizer): distinct agg.

### PR DESCRIPTION
**Pull request by likg227**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)

  I add DistinctAgg rule to transform distinct aggregates according to [Eric's comment](https://github.com/singularity-data/risingwave/issues/146#issuecomment-1162712210).

- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

  Because the backend implementation of selective aggregates is still in progress, no e2e test can be added now.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

The rule will convert pattern `agg -> input` to pattern `agg -> agg -> expand -> input`.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Plan change.

## Refer to a related PR or issue link (optional)

related #3589

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 1]
```diff
@@ -124,6 +124,10 @@ impl PlanAggCall {
             filter: Condition::true_cond(),
         }
     }
+
+    pub fn input_indcies(&self) -> Vec<usize> {
```
Comment (xiangjinwu) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:128]:
```suggestion
    pub fn input_indices(&self) -> Vec<usize> {
```
typo

File: src/frontend/src/optimizer/rule/distinct_agg.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+///
+/// will be transformed to
+///
+/// `LogicalAgg(group by(0, 1), sum($4) filter(where $3 = 1), sum($5) filter(where $3 = 0))`
+///
+/// -> `LogicalAgg(group by(0, 1, 3, 4), count($2) filter(where $2 < 100), sum($3))`
+///
+/// -> `Expand(column_subsets: [[0, 1, 3], [0, 1, 2, 2]])` -> `Input(len of schema: 4)`
+pub struct DistinctAgg {}
+impl Rule for DistinctAgg {
```
Comment (st1page) [inline src/frontend/src/optimizer/rule/distinct_agg.rs:40]:
```suggestion
impl Rule for DistinctAggToExpand {
```

**Review by st1page — APPROVED** [review]
rest LGTM

**Comment by chenzl25** [comment]
Try this case select count(*) as cnt, count(distinct v2) from t.
Make sure that when t is an empty table, cnt can return 0.
By the way, what's the difference between AggKind::COUNT and AggKind::ROW_COUNT.

**Comment by likg227** [comment]
> Try this case select count(*) as cnt, count(distinct v2) from t.
> Make sure that when t is an empty table, cnt can return 0.

Because selective aggregates is not fully supported, I can't add e2e tests now and just planner test is available. I will add this test when selective aggregates is ready.

> By the way, what's the difference between AggKind::COUNT and AggKind::ROW_COUNT.

IIUC, AggKind::COUNT is used to denote normal count(e.g. count(v1), while AggKind::ROW_COUNT was used to denote special count(i.e. count(*)).  But we do not use AggKind::ROW_COUNT to denote count(*) anymore(AggKind::COUNT with zero argument instead). cc @skyzh

File: src/frontend/test_runner/tests/testdata/agg.yaml [hunk 3]
```diff
@@ -496,3 +496,42 @@
...
+- sql: |
+    /* distinct agg and non-disintct agg */
+    create table t(a int, b int, c int);
+    select a, count(distinct b) as distinct_b_num, sum(c) as sum_c from t group by a;
+  optimized_logical_plan: |
+    LogicalAgg { group_key: [0], agg_calls: [sum($2) filter(($1 = 0:Int64)), sum($3) filter(($1 = 1:Int64))] }
+      LogicalProject { exprs: [$0, $2, $3, $4] }
+        LogicalAgg { group_key: [0, 1, 3], agg_calls: [count($1), sum($2)] }
+          LogicalExpand { column_subsets: [[0, 1], [0, 2]] }
+            LogicalScan { table: t, columns: [a, b, c] }
```
Comment (xiangjinwu) [inline src/frontend/test_runner/tests/testdata/agg.yaml:508]: This plan looks problematic:
Assume `t` has the following rows:
```
1 2 3
1 2 4
```
The expected output is:
```
1 1 7
```

Without the ability to actually run it e2e, let's execute it in mind:
After expand:
```
1 2 N 0
1 2 N 0
1 N 3 1
1 N 4 1
```
Then `LogicalAgg { group_key: [0, 1, 3], agg_calls: [count($1), sum($2)] }`:
```
1 2 0 2 N
1 N 1 0 7
```
Then project:
```
1 0 2 N
1 1 0 7
```
Finally:
```
1 2 7
```

The error **in this case** is that we `count` when group by `a, b, flag`, then `sum` that result. However, The 1st `count` is **unnecessary**, we just need to `group by a, b, flag` and then `count` the `b` from group column. This also aligns with the original comment in issue.

Intuitively, for an agg-distinct call, the earlier agg node does `distinct`, and the later agg node does the real `agg`. At the same time, for an agg without distinct, the earlier agg node does the real `agg`, and the later agg node may be able to use `single_value` (correct me if I am wrong).

---

The "fix" above cannot handle the case `count(distinct b) filter (where d > 5)`, where `d` is neither a group column or used by any other agg calls. We will have to `filter d > 5` before we can `distinct b` via `group by b` of the earlier agg node. Some premature ideas I just come up with:
* `count(*) filter (where d > 5)` in earlier agg node, then `filter(where flag = x and count_star > 0)`. Note that it is always `count` no matter what the original agg call is. This feels like a bit wasteful.
* add a filter node between expand and earlier agg node

File: src/frontend/test_runner/tests/testdata/agg.yaml [hunk 4]
```diff
@@ -496,3 +496,44 @@
...
+- sql: |
+    /* non-distinct agg with filter */
+    create table t(a int, b int, c int);
+    select a, count(distinct b), sum(c) filter(where b < 100) from t group by a;
+  optimized_logical_plan: |
+    LogicalAgg { group_key: [0], agg_calls: [count($1) filter(($2 = 1:Int64)), sum($3) filter(($2 = 0:Int64))] }
+      LogicalAgg { group_key: [0, 1, 3], agg_calls: [sum($2) filter(($1 < 100:Int32))] }
+        LogicalExpand { column_subsets: [[0, 1, 2], [0, 1]] }
+          LogicalScan { table: t, columns: [a, b, c] }
+- sql: "/* only distinct agg */ \ncreate table t(a int, b int, c int); \nselect a, count(distinct b) as distinct_b_num, sum(distinct c) filter(where c < 100) as distinct_c_sum from t group by a;\n"
```
Comment (xiangjinwu) [inline src/frontend/test_runner/tests/testdata/agg.yaml:534]: try format as multiple lines?

File: src/frontend/test_runner/tests/testdata/tpch.yaml [hunk 5]
```diff
@@ -1661,48 +1661,56 @@
...
-    LogicalAgg { group_key: [0, 1, 2], agg_calls: [count(distinct $3)] }
-      LogicalProject { exprs: [$1, $2, $3, $0] }
-        LogicalJoin { type: LeftAnti, on: ($0 = $4), output_indices: all }
-          LogicalJoin { type: Inner, on: ($2 = $0), output_indices: [1, 3, 4, 5] }
-            LogicalScan { table: partsupp, columns: [ps_partkey, ps_suppkey] }
-            LogicalScan { table: part, output_columns: [p_partkey, p_brand, p_type, p_size], required_columns: [$1:p_partkey, $4:p_brand, $5:p_type, $6:p_size], predicate: ($4 <> 'Brand#45':Varchar) AND Not(Like($5, 'SMALL PLATED%':Varchar)) AND In($6, 19:Int32, 17:Int32, 16:Int32, 23:Int32, 10:Int32, 4:Int32, 38:Int32, 11:Int32) }
-          LogicalScan { table: supplier, output_columns: [s_suppkey], required_columns: [$1:s_suppkey, $7:s_comment], predicate: Like($7, '%Customer%Complaints%':Varchar) }
+    LogicalAgg { group_key: [0, 1, 2], agg_calls: [count($3) filter(($4 = 0:Int64))] }
+      LogicalAgg { group_key: [0, 1, 2, 3, 4], agg_calls: [] }
+        LogicalExpand { column_subsets: [[0, 1, 2, 3]] }
```
Comment (xiangjinwu) [inline src/frontend/test_runner/tests/testdata/tpch.yaml:1666]: Maybe open an issue to simplify / remove unnecessary expand.

File: src/frontend/src/optimizer/rule/distinct_agg.rs [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                index_of_middle_agg += 1;
+
+                // the filter of non-distinct agg has been calculated in middle agg.
+                agg_call.filter = Condition::true_cond();
+
+                // change final agg's agg_kind just like two-phase agg.
+                match agg_call.agg_kind {
+                    AggKind::Count
+                    | AggKind::RowCount
+                    | AggKind::Sum
```
Comment (xiangjinwu) [inline src/frontend/src/optimizer/rule/distinct_agg.rs:180]: * rebase and there will be no `RowCount`
* it is okay to omit `Sum` here

File: src/frontend/test_runner/tests/testdata/tpch.yaml (hunk 5 above)
Comment (likg227) [reply src/frontend/test_runner/tests/testdata/tpch.yaml:1669]: #3806

**Review by xiangjinwu — APPROVED** [review]

**Comment by xiangjinwu** [comment]
As sync'ed offline, the following special cases will be handled in upcoming PR:
* `count` returns 0 rather than null in https://github.com/singularity-data/risingwave/issues/3717#issuecomment-1178597987
* `agg0(distinct)` and `agg1(order by)` cannot be in the same query.

**Pull request merged**